src/core/mac/mac_types.hpp

```diff
@@ -599,8 +600,57 @@ class NetworkName : public otNetworkName
      * @retval OT_ERROR_INVALID_ARGS   Given name is too long.
      *
      */
-    otError Set(const Data &aNameData);
+    otError Set(const NameData &aNameData);
+};
+
+#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
+/**
+ * This structure represents a Thread Domain Name.
+ *
+ */
+class DomainName : public otDomainName
+{
+public:
+    enum
+    {
+        kMaxSize = OT_DOMAIN_NAME_MAX_SIZE, // Maximum number of chars in Domain Name (excludes null char).
+    };
+
+    /**
+     * This constructor initializes the Thread Domain Name as an empty string.
+     *
+     */
+    DomainName(void) { m8[0] = '\0'; }
+
+    /**
+     * This method gets the Thread Domain Name as a null terminated C string.
+     *
+     * @returns The Domain Name as a null terminated C string array.
+     *
+     */
+    const char *GetAsCString(void) const { return m8; }
+
+    /**
+     * This method gets the Thread Domain Name as NameData.
+     *
+     * @returns The Domain Name as NameData.
+     *
+     */
+    NameData GetAsData(void) const;
+
+    /**
+     * This method sets the Thread Domain Name.
+     *
+     * @param[in]  aNameData           A reference to name data.
+     *
+     * @retval OT_ERROR_NONE           Successfully set the Thread Domain Name.
+     * @retval OT_ERROR_ALREADY        The name is already set to the same string.
+     * @retval OT_ERROR_INVALID_ARGS   Given name is too long.
+     *
+     */
+    otError Set(const NameData &aNameData);
```

Comment:
Question: Will we have use-cases later that would need to treat the `DomainName` as a `Data` (e.g. when attaching to a TLV, etc).

Reply:
It will be included in Discovery Response for Commercial Commissioning Extensions feature someday.  
Update as what this PR is out of considerations the similarity with NetworkName in some extent (now the naming, future the tlv parsing).